istioctl/cmd/vm-bootstrap.go

```diff
@@ -0,0 +1,630 @@
+		}
+		sshAuthMethod = ssh.Password(sshPassword)
+	} else {
+		// Attempt to parse the key.
+		rawKey, err := ioutil.ReadFile(sshKeyLocation)
+		if err != nil {
+			return err
+		}
+		key, err := ssh.ParsePrivateKey(rawKey)
+		if err != nil {
+			if err, ok := err.(*ssh.PassphraseMissingError); ok {
+				term := terminal.NewTerminal(os.Stdin, "")
+				sshKeyPassword, err := term.ReadPassword("Please enter the password for the ssh key: ")
+				if err != nil {
+					return err
+				}
+				decryptedKey, err := ssh.ParsePrivateKeyWithPassphrase(rawKey, []byte(sshKeyPassword))
+				if err != nil {
+					return err
+				}
+				sshAuthMethod = ssh.PublicKeys(decryptedKey)
+			} else {
+				return err
+			}
+		} else {
+			sshAuthMethod = ssh.PublicKeys(key)
+		}
+	}
+
+	return nil
+}
+
+func vmBootstrapCommand() *cobra.Command {
+	vmBSCommand := &cobra.Command{
+		Use:   "vm-bootstrap",
+		Short: "bootstraps a workloadentry representing a vm into the mesh",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if (len(args) == 1) == all {
+				cmd.Println(cmd.UsageString())
+				return fmt.Errorf("vm-bootstrap requires a workload entry, or the all flag")
+			}
+			if all && workloadNamespace == "" {
+				return fmt.Errorf("vm-bootstrap needs a namespace if fetching all workspaces")
+			}
+			if sshUser == "" {
+				user, err := user.Current()
+				if err != nil {
+					return err
+				}
+				sshUser = user.Username
+			}
+			if dumpDir == "" {
+				err := deriveSSHMethod()
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+		RunE: func(c *cobra.Command, args []string) error {
+			var configClient model.ConfigStore
+			var err error
+
+			if configClient, err = clientFactory(); err != nil {
+				return err
+			}
+
+			var entries []model.Config
+			var chosenNS string
+			if all {
+				entries, chosenNS, err = fetchAllWorkloadEntries(configClient)
+			} else {
+				entries, chosenNS, err = fetchSingleWorkloadEntry(args[0], configClient)
+			}
+			if err != nil {
+				return err
+			}
+
+			kubeClient, err := interfaceFactory(kubeconfig)
+			if err != nil {
+				return err
+			}
+
+			certs, err := getCertificate(kubeClient)
+			if err != nil {
+				return err
+			}
+
+			addresses, err := getCertificatesForEachAddress(entries, chosenNS, certs)
+			if err != nil {
+				return err
+			}
+
+			if dumpDir != "" {
+				err = dumpCertificates(dumpDir, addresses)
+				if err != nil {
+					return err
+				}
+			} else {
+				err = copyCertificates(kubeClient, addresses)
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+	}
+
+	vmBSCommand.PersistentFlags().BoolVarP(&all, "all", "a", false,
+		"attempt to bootstrap all workload entries")
+	vmBSCommand.PersistentFlags().StringVarP(&dumpDir, "dump-directory", "d", "",
+		"dump the certificates locally in the provided directory instead of uploading remotely.")
+	vmBSCommand.PersistentFlags().DurationVar(&certDuration, "duration", 365*24*time.Hour,
```

Comment:
Is a year the best default?  Maybe shorter, and require the user to explicitly provide an arg for such a long duration?

Reply:
The idea was specifically long lived certificates _for now_, in order to combat say a VM rebooting, and needing to bootstrap an identity again. Since it'd be awkward to have a VM Reboot, and need to be re-bootstrapped.